src/codecs/framing/character_delimited.rs

```diff
@@ -87,80 +81,53 @@ impl Decoder for CharacterDelimitedDecoder {
+                            buf_len = buf.len(),
+                            max_length = self.max_length,
+                            internal_log_rate_secs = 30
+                        );
+                        buf.advance(next_delimiter_idx + 1);
+                    } else {
+                        let frame = buf.split_to(next_delimiter_idx).freeze();
+                        trace!(
+                            message = "Decoding the frame.",
+                            bytes_proccesed = frame.len()
+                        );
+                        buf.advance(1); // scoot past the delimiter
+                        return Ok(Some(frame));
                     }
                 }
-                (false, Some(pos)) => {
-                    // We found a correct frame
-
-                    let newpos_index = pos + self.next_index;
-                    self.next_index = 0;
-                    let mut frame = buf.split_to(newpos_index + 1);
-
-                    trace!(
-                        message = "Decoding the frame.",
-                        bytes_proccesed = frame.len()
-                    );
-
-                    let frame = frame.split_to(frame.len() - 1);
-
-                    return Ok(Some(frame.freeze()));
-                }
-                (false, None) if buf.len() > self.max_length => {
-                    // We reached the max length without finding the
-                    // delimiter so must discard the rest until we
-                    // reach the next delimiter
-                    self.is_discarding = true;
-                    warn!(
-                        message = "Discarding frame larger than max_length.",
-                        buf_len = buf.len(),
-                        max_length = self.max_length,
-                        internal_log_rate_secs = 30
-                    );
-                }
-                (false, None) => {
-                    // We didn't find the delimiter and didn't
-                    // reach the max frame length.
-                    self.next_index = read_to;
-                    return Ok(None);
-                }
             }
         }
     }
 
     fn decode_eof(&mut self, buf: &mut BytesMut) -> Result<Option<Bytes>, Self::Error> {
-        let frame = self.decode(buf)?.or_else(|| {
-            (!buf.is_empty() && !self.is_discarding).then(|| {
-                self.next_index = 0;
-                buf.split_to(buf.len()).into()
-            })
-        });
-
-        Ok(frame)
+        match self.decode(buf)? {
+            Some(frame) => Ok(Some(frame)),
+            None => {
+                if buf.is_empty() {
+                    Ok(None)
+                } else {
+                    // TODO why do we do this? We drop all but the final byte of
+                    // the incoming buffer?
+                    let bytes: Bytes = buf.split_to(buf.len()).freeze();
```

Comment:
I believe the intent here is to, if an EOF is found before finding the delimiter, to just return the remainder as a frame. I think we need the length check here too though. Previously, the check for `is_discarding` would have avoided returning frames that were too large if the delimiter wasn't found before EOF, but now I think this implementation would. It could be good to add a test for this behavior.

Reply:
Ah, interesting. So `split_to` doesn't include the last byte so if the intend was to return the remainder of the frame this was buggy: it returned the remainder less the final byte.

Reply:
Ah, you are right, that does seem like a bug then. It should just return the remainder. Again, [using `LinesCodec` as a reference](https://docs.rs/tokio-codec/latest/src/tokio_codec/lines_codec.rs.html#166-183) they just do `.take()`.

Reply:
Ah, I see. It's not totally straightforward to return the entire buffer considering we don't actually own the buffer. I'll open a follow-up issue for the dropped-byte since that's not an introduced here but will add the max_length guard.

Reply:
🤔 can't we just call `.take()` instead of `split_to()`? https://docs.rs/bytes/1.1.0/bytes/buf/trait.Buf.html#method.take

Either way, yeah, that's fine, we can follow-up on the dropped byte.

Reply:
That gives you a `Take` but the interior is still a `BytesMut`. I'm sure there's some chain of calls to get what we want, but it's not clear to me right now.

Reply:
https://github.com/vectordotdev/vector/issues/10583

Reply:
Actually, I think `split_to` is fine?

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=3abdbba7c55459d111ed67d0f26c0e7e

Reply:
Ah, off-by-one error in my thinking around this. I miscalculated: len is the number of bytes not the last index. Not the first time I've made that mistake.

Reply:
You and me both.